nsxt/data_source_nsxt_edge_upgrade_group.go

```diff
@@ -0,0 +1,144 @@
+/* Copyright © 2024 VMware, Inc. All Rights Reserved.
+   SPDX-License-Identifier: MPL-2.0 */
+
+package nsxt
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt-mp/nsx/model"
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt-mp/nsx/upgrade"
+)
+
+var (
+	edgeUpgradeGroup = "EDGE"
+	hostUpgradeGroup = "HOST"
+
+	timeoutUUGCreate  = 1 * time.Minute
+	intervalUUGCreate = 5 * time.Second
+)
+
+func dataSourceNsxtEdgeUpgradeGroup() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceNsxtEdgeUpgradeGroupRead,
+
+		Schema: map[string]*schema.Schema{
+			"upgrade_prepare_id": {
+				Type:        schema.TypeString,
+				Description: "ID of corresponding nsxt_upgrade_prepare resource",
+				Required:    true,
+			},
+			"id":           getDataSourceIDSchema(),
+			"display_name": getDataSourceDisplayNameSchema(),
+			"description":  getDataSourceDescriptionSchema(),
+		},
+	}
+}
+
+func dataSourceNsxtEdgeUpgradeGroupRead(d *schema.ResourceData, m interface{}) error {
+	return upgradeGroupRead(d, m, edgeUpgradeGroup)
+}
+
+func upgradeGroupRead(d *schema.ResourceData, m interface{}, groupType string) error {
+	connector := getPolicyConnector(m)
+	client := upgrade.NewUpgradeUnitGroupsClient(connector)
+
+	err := waitUpgradeGroupCreate(client, groupType)
```

Comment:
This would be an extra roundtrip for each group.. I don't think it would be very noticeable, but I'm also not sure this is needed one we have the `upgrade_prepare_id` being forced on the user. The prepare resource waits for quite a few async processes, and I would expect the groups to be created way before it finishes - do you think further wait would be required here?
I would suggest to remove this wait for now, and add it only if we experience the actual need for it. What do you think?

Reply:
Make sense. And based on my test those groups will be created in the early stage of uploading the upgrade bundle. We could add it back later if necessary.